From 1edffc9a491740e44e02efccb82735cbc4c5411d Mon Sep 17 00:00:00 2001
From: Lubomir Rintel <lkundrak@v3.sk>
Date: Sun, 18 Nov 2018 23:01:13 +0100
Subject: [PATCH] imap_proxy: spin for 10 seconds for the IMAP proxy socket to
 appear

While far from ideal (why don't we need a socket instead of piping things
via stdin/stdout?), this fixes a timeout on slower machines, while
potentially improving startup time on newer ones.
---
 src/imap_proxy.cc | 30 ++++++++++++++++++++++--------
 src/imap_proxy.h  |  5 +++++
 2 files changed, 27 insertions(+), 8 deletions(-)

diff --git a/src/imap_proxy.cc b/src/imap_proxy.cc
index 7c9c020..65d8945 100644
--- a/src/imap_proxy.cc
+++ b/src/imap_proxy.cc
@@ -40,6 +40,12 @@
 CIMAPProxy::CIMAPProxy()
 {
     m_child = -1;
+
+    /*
+     * Use ~/.imap.sock as the path.
+     */
+    m_sock_path = getenv("HOME");
+    m_sock_path += "/.imap.sock";
 }
 
 
@@ -92,8 +98,11 @@ void CIMAPProxy::launch()
         if (CFile::exists(path))
         {
             CStatusPanel *panel = CStatusPanel::instance();
+            int i;
+
             panel->add_text("Launching IMAP proxy " + path);
 
+            unlink(m_sock_path.c_str());
             m_child = fork();
 
             if (m_child == 0)
@@ -102,7 +111,18 @@ void CIMAPProxy::launch()
                 exit(1);
             }
 
-            sleep(1.0);
+            for (i = 0; i < 100; i++)
+            {
+                if (access(m_sock_path.c_str(), F_OK) == 0)
+                {
+                    /*
+                     * Done waiting, successfully.
+                     */
+                    return;
+                }
+                usleep(100000);
+            }
+            panel->add_text("Timed out waiting for IMAP proxy");
         }
         else
         {
@@ -134,13 +154,7 @@ std::string CIMAPProxy::read_imap_output(std::string cmd)
     sockfd = socket(AF_UNIX, SOCK_STREAM, 0);
     addr.sun_family = AF_UNIX;
 
-    /*
-     * Use ~/.imap.sock as the path.
-     */
-    std::string path = getenv("HOME");
-    path += "/.imap.sock";
-
-    strcpy(addr.sun_path, path.c_str());
+    strcpy(addr.sun_path, m_sock_path.c_str());
 
     if (connect(sockfd, (sockaddr*)&addr, sizeof(addr)) < 0)
     {
diff --git a/src/imap_proxy.h b/src/imap_proxy.h
index a1acf21..76808bb 100644
--- a/src/imap_proxy.h
+++ b/src/imap_proxy.h
@@ -64,4 +64,9 @@ class CIMAPProxy : public Singleton<CIMAPProxy>
      * The handle to our child-process.
      */
     pid_t m_child;
+
+    /**
+     * Path to the IMAP proxy socket.
+     */
+    std::string m_sock_path;
 };
